R/tag_query.R

```diff
@@ -0,0 +1,1601 @@
+      count <<- count + 1
+      if (count == 500) {
+        # convert the current stack to a `envirStackUni_()`
+        newStack <- envirStackUni_()
+        walk(stack$asList(), newStack$push)
+        stack <<- newStack
+      }
+      stack$push(env)
+    },
+    uniqueList = function() {stack$uniqueList()}
+  )
+}
+
+
+
+
+
+
+# Retrieve all attributes that can be manually set
+# ?attr
+# Note that some attributes (namely ‘class’, ‘comment’, ‘dim’,
+# ‘dimnames’, ‘names’, ‘row.names’ and ‘tsp’) are treated specially
+# and have restrictions on the values which can be set.
+safeAttrValues <- function(x) {
+  badElAttrs <- c("class", "comment", "dim", "dimnames", "names", "row.names", "tsp")
+  attrVals <- attributes(x)
+  attrVals[badElAttrs] <- NULL
+  attrVals
+}
+
+# Convert a list to an environment and keep class and attribute information
+safeListToEnv <- function(x, newClass = NULL) {
+  xList <- x
+
+  ret <- list2env(xList, new.env(parent = emptyenv()))
+
+  attrVals <- safeAttrValues(xList)
+  walk2(names(attrVals), attrVals, function(attrName, attrValue) {
+    attr(ret, attrName) <<- attrValue
+  })
+
+  class(ret) <- c(newClass, class(xList), "environment")
+  ret
+}
+
+# Convert an environment to a list and keep class and attribute information
+safeEnvToList <- function(x, oldClass = NULL) {
+  xEnv <- x
+  ret <- as.list.environment(xEnv, all.names = TRUE)
+
+  attrVals <- safeAttrValues(xEnv)
+  walk2(names(attrVals), attrVals, function(attrName, attrValue) {
+    attr(ret, attrName) <<- attrValue
+  })
+
+  class(ret) <- setdiff(class(x), c(oldClass, "environment"))
+  ret
+}
+
+
+# Do not export to encourage direct use of `tagQuery()`
+asTagEnv <- function(x) {
+  if (isTagQuery(x)) {
+    return(asTagEnv(x$root()))
+  }
+
+  if (!isTagEnv(x)) {
+    if (!isTag(x)) {
+      # force all methods to send in tags, lists / tagLists are not allowed
+      stop("`asTagEnv()` can only accept tag envs or tag objects. It does not accept `lists()` or `tagLists()`")
+    }
+  }
+  asTagEnv_(x, parent = x$parent)
+}
+asTagEnv_ <- function(x, parent = NULL, seenMap = envirMap()) {
+  isTagVal <- isTag(x)
+  isTagEnvVal <- isTagEnv(x)
+
+  if (isTagVal || isTagEnvVal) {
+    if (!isTagEnvVal) {
+      xList <- x
+      x <- safeListToEnv(xList, "htmltools.tag.env")
+      # add parent env and key
+      x$parent <- parent
+      x$envKey <- sexp_address(x)
+    }
+    if (seenMap$has(x)) {
+      stop(
+        "Circular family tree found with tag environment: ", sexp_address(x), "\n",
+        # Not necessarily the order of the circular dependency
+        # TODO-later show actual circular dependency and not all visited nodes? This should be rare
+        "Tags processed: (unordered set)\n", paste0("* ", seenMap$keys(), collapse = "\n")
+      )
+    }
+    # Add the item to the seen map to help with cycle detection
+    seenMap$set(x, TRUE)
+
+    # Make sure all attribs are unique
+    x$attribs <- flattenTagAttribs(x$attribs)
```

Comment:
We should be able to get rid of this once we implement flattening attributes on the way in.

Reply:
... This is the way in?

Reply:
See #205.

Reply:
Will add in another PR. (I think this will be breaking behavior)